drivers/rtt_rtc/rtt_rtc.c

```diff
@@ -137,28 +138,35 @@ int rtc_set_time(struct tm *time)
 
 int rtc_get_time_ms(struct tm *time, uint16_t *ms)
 {
-    uint32_t prev = rtc_now;
+    uint32_t tmp, prev;
 
     /* repeat calculation if an alarm triggered in between */
     do {
-        uint32_t now = rtt_get_counter();
-        uint32_t tmp = _rtc_now(now);
+        uint32_t now;
 
-        rtc_localtime(tmp, time);
-        *ms = (SUBSECONDS(now) * MS_PER_SEC) / RTT_SECOND;
+        prev = rtc_now;
+        now  = rtt_get_counter();
+        tmp  = _rtc_now(now);
+
+        *ms = (SUBSECONDS(now - last_alarm) * MS_PER_SEC) / RTT_SECOND;
     } while (prev != rtc_now);
```

Comment:
OK, let's assume that `last_alarm` would be `volatile` here as well and we are on an AVR with a 16 bit RTT.

1. `rtc_now` has a value of `0x00ff`
2. `prev = rtc_now;` requires two 8 bit loads and in between these loads `rtc_now` is set to `0x4200`. As a result, `prev` has a value of `0x42ff`
3. The read of `last_alarm` gets similarly corrupted by the value changing in between the reads of the two least significant bytes
4. The timer fires and `rtc_now` gets a value of `0x42ff`
5. `while (prev != rtc_now)` doesn't detect the corruptions, as `prev` has a corrupted value of `0x42ff`

I agree that the chance of something like this happening is pretty low. But this only makes the issue harder to debug.